# Exploratory Notebook

### Exploring the Data

importing pandas with its alias pd

In [1]:
import pandas as pd

Unzipping the zipped data files (all at once)

In [15]:
#!find . -name '*.tsv.gz' -exec gzip -d {} \;

Reading in all the data files to understand which ones will help answer our business questions. Checking the files with .head() and .tail() methods to view a section of the data. The .shape method will help tell how many rows we have in each data set. The .info() method will help discover the dtypes of the columns and possible missing values.

In [5]:
df0= pd.read_csv('./data/zippedData/imdb.name.basics.csv')
df0.head()
#df0.shape

nconst       primary_name  birth_year  death_year  \
0  nm0061671  Mary Ellen Bauder         NaN         NaN   
1  nm0061865       Joseph Bauer         NaN         NaN   
2  nm0062070         Bruce Baum         NaN         NaN   
3  nm0062195       Axel Baumann         NaN         NaN   
4  nm0062798        Pete Baxter         NaN         NaN   

                                 primary_profession  \
0         miscellaneous,production_manager,producer   
1        composer,music_department,sound_department   
2                        miscellaneous,actor,writer   
3  camera_department,cinematographer,art_department   
4  production_designer,art_department,set_decorator   

                          known_for_titles  
0  tt0837562,tt2398241,tt0844471,tt0118553  
1  tt0896534,tt6791238,tt0287072,tt1682940  
2  tt1470654,tt0363631,tt0104030,tt0102898  
3  tt0114371,tt2004304,tt1618448,tt1224387  
4  tt0452644,tt0452692,tt3458030,tt2178256

In [31]:
df1= pd.read_csv('./data/zippedData/bom.movie_gross.csv')
df1.head()

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [7]:
df2= pd.read_csv('./data/zippedData/imdb.title.akas.csv')
df2.head()

title_id  ordering                                    title region  \
0  tt0369610        10                            Джурасик свят     BG   
1  tt0369610        11                        Jurashikku warudo     JP   
2  tt0369610        12  Jurassic World: O Mundo dos Dinossauros     BR   
3  tt0369610        13                  O Mundo dos Dinossauros     BR   
4  tt0369610        14                           Jurassic World     FR   

  language        types   attributes  is_original_title  
0       bg          NaN          NaN                0.0  
1      NaN  imdbDisplay          NaN                0.0  
2      NaN  imdbDisplay          NaN                0.0  
3      NaN          NaN  short title                0.0  
4      NaN  imdbDisplay          NaN                0.0

The dataset below gives us important information about the movies. Important columns: primary_title, start_year, and genres.

In [44]:
df3= pd.read_csv('./data/zippedData/imdb.title.basics.csv')
df3.head()
df_title_movie = df3.rename(columns = {'primary_title':'movie'})
df_title_movie.head()

tconst                            movie              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

Changed the primary_title to 'movie' so that the merge with the next important data set goes smoothly. 

In [92]:
df_title_movie.describe()

start_year  runtime_minutes
count  146144.000000    114405.000000
mean     2014.621798        86.187247
std         2.733583       166.360590
min      2010.000000         1.000000
25%      2012.000000        70.000000
50%      2015.000000        87.000000
75%      2017.000000        99.000000
max      2115.000000     51420.000000

In [9]:
df4= pd.read_csv('./data/zippedData/imdb.title.crew.csv')
df4.head()

tconst                      directors              writers
0  tt0285252                      nm0899854            nm0899854
1  tt0438973                            NaN  nm0175726,nm1802864
2  tt0462036                      nm1940585            nm1940585
3  tt0835418                      nm0151540  nm0310087,nm0841532
4  tt0878654  nm0089502,nm2291498,nm2292011            nm0284943

In [10]:
df5= pd.read_csv('./data/zippedData/imdb.title.principals.csv')
df5.head()

tconst  ordering     nconst  category       job        characters
0  tt0111414         1  nm0246005     actor       NaN       ["The Man"]
1  tt0111414         2  nm0398271  director       NaN               NaN
2  tt0111414         3  nm3739909  producer  producer               NaN
3  tt0323808        10  nm0059247    editor       NaN               NaN
4  tt0323808         1  nm3579312   actress       NaN  ["Beth Boothby"]

In [11]:
df6= pd.read_csv('./data/zippedData/imdb.title.ratings.csv')
df6.head()

tconst  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [12]:
df7= pd.read_csv('./data/zippedData/tmdb.movies.csv')
df7.head()

Unnamed: 0            genre_ids     id original_language  \
0           0      [12, 14, 10751]  12444                en   
1           1  [14, 12, 16, 10751]  10191                en   
2           2        [12, 28, 878]  10138                en   
3           3      [16, 35, 10751]    862                en   
4           4        [28, 878, 12]  27205                en   

                                 original_title  popularity release_date  \
0  Harry Potter and the Deathly Hallows: Part 1      33.533   2010-11-19   
1                      How to Train Your Dragon      28.734   2010-03-26   
2                                    Iron Man 2      28.515   2010-05-07   
3                                     Toy Story      28.005   1995-11-22   
4                                     Inception      27.920   2010-07-16   

                                          title  vote_average  vote_count  
0  Harry Potter and the Deathly Hallows: Part 1           7.7       10788  
1                      How to Train Your Dragon           7.7        7610  
2                                    Iron Man 2           6.8       12368  
3                                     Toy Story           7.9       10174  
4                                     Inception           8.3       22186

Found the next file to be one of the important files to gain an understanding of the costs and profits of different movies.

In [73]:
df8= pd.read_csv('./data/zippedData/tn.movie_budgets.csv')
df8['start_year']= [int(x[-4:]) for x in df8['release_date']]
df8.head()
#df8.loc[df8['domestic_gross'] == '$0']

id  release_date                                        movie  \
0   1  Dec 18, 2009                                       Avatar   
1   2  May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2   3   Jun 7, 2019                                 Dark Phoenix   
3   4   May 1, 2015                      Avengers: Age of Ultron   
4   5  Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   

  production_budget domestic_gross worldwide_gross  start_year  
0      $425,000,000   $760,507,625  $2,776,345,279        2009  
1      $410,600,000   $241,063,875  $1,045,663,875        2011  
2      $350,000,000    $42,762,350    $149,762,350        2019  
3      $330,600,000   $459,005,868  $1,403,013,963        2015  
4      $317,000,000   $620,181,382  $1,316,721,747        2017

Created a new column called start_year that creates a year integer from the string from release_date. This is another qualifier for merging this dataset and the one chosen from above.

### Cleaning the Data

First, merging the datasets so that we have all the information we need in one dataframe.
    The 'inner' join is chosen because each dataset has important data for the analysis. Not having both sides of the dataset makes the row useless in our chosen analysis. The merge is on 'movie' and 'start_year' to make sure that duplicates are at a minimum from having movies with the same name, but different release years.
    We also check for duplicates and drop the rows from our merged dataset.

In [80]:
df_budget_merge = pd.merge(df8, df_title_movie, how = 'inner', on = ('movie', 'start_year'))
df_duplicates = df_budget_merge[df_budget_merge['movie'].duplicated()]
#df_budget_merge.loc[df_budget_merge['domestic_gross'] == '$0'].head()
df_budget_merge.drop(df_duplicates.index, axis = 0, inplace = True)
#df_budget_merge.head()

64


We next check for missing values. We use worlwide_gross as a check to see if any data necessary for profit calculation is missing. If it is, we drop the row because filling the missing values would throw off our analysis and it does not account for more recently, stream only movies like from Netflix. We want to check only movies from the box office.

In [97]:
df_no_values = df_budget_merge.loc[df_budget_merge['worldwide_gross'] == '$0']
df_budget_merge.drop(df_no_values.index, axis = 0, inplace = True)

Below is a function to change any string numbers into integers so we can use them for stats calculation.

In [102]:
def clean_dollars(dataframe, column_str):
    dataframe[column_str] = dataframe[column_str].str.replace(',', '').str.replace('$', '').astype(int)
    return dataframe

#clean_dollars(df_budget_merge, 'production_budget')
clean_dollars(df_budget_merge, 'domestic_gross')
clean_dollars(df_budget_merge, 'worldwide_gross')
df_budget_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1365 entries, 0 to 1542
Data columns (total 11 columns):
id                   1365 non-null int64
release_date         1365 non-null object
movie                1365 non-null object
production_budget    1365 non-null int64
domestic_gross       1365 non-null int64
worldwide_gross      1365 non-null int64
start_year           1365 non-null int64
tconst               1365 non-null object
original_title       1365 non-null object
runtime_minutes      1359 non-null float64
genres               1365 non-null object
dtypes: float64(1), int64(5), object(5)
memory usage: 128.0+ KB


Next we create an advertisement column to add another important cost to producing a movie. Due to not having the data, we chose to follow the heuristic that advertisement budgets will generally be equal to production budgets. With the costs, we create a profit column from the differnce between the worldwide_gross column and the production_budget and advertisement_budget columns.

In [124]:
df_budget_merge['advertisement_budget'] = df_budget_merge['production_budget']
df_budget_merge['profit'] = df_budget_merge['worldwide_gross'] - df_budget_merge['production_budget'] - df_budget_merge['advertisement_budget']
df_budget_merge.head(20)

id  release_date                           movie  production_budget  \
3     7  Apr 27, 2018          Avengers: Infinity War          300000000   
23   34  Jun 12, 2015                  Jurassic World          215000000   
46   67   Apr 3, 2015                       Furious 7          190000000   
18   27   May 4, 2012                    The Avengers          225000000   
301  73  Jul 10, 2015                         Minions           74000000   
108  56  Nov 22, 2013                          Frozen          150000000   
77   13  Jun 22, 2018  Jurassic World: Fallen Kingdom          170000000   
27   42  Feb 16, 2018                   Black Panther          200000000   
95   35  Mar 17, 2017            Beauty and the Beast          160000000   
293  30  Jun 30, 2017                 Despicable Me 3           75000000   
28   44  Jun 15, 2018                   Incredibles 2          200000000   
96   36  Dec 21, 2018                         Aquaman          160000000   
290  22   Jul 3, 2013                 Despicable Me 2           76000000   
32   48   May 3, 2013                      Iron Man 3          200000000   
383  84   Nov 2, 2018               Bohemian Rhapsody           55000000   
240  38  Dec 20, 2017  Jumanji: Welcome to the Jungle           90000000   
67   96   Mar 8, 2019                  Captain Marvel          175000000   
2     4   May 1, 2015         Avengers: Age of Ultron          330600000   
291  26   Jul 8, 2016         The Secret Life of Pets           75000000   
16   23  Apr 14, 2017         The Fate of the Furious          250000000   

     domestic_gross  worldwide_gross  start_year     tconst  \
3         678815482       2048134200        2018  tt4154756   
23        652270625       1648854864        2015  tt0369610   
46        353007020       1518722794        2015  tt2820852   
18        623279547       1517935897        2012  tt0848228   
301       336045770       1160336173        2015  tt2293640   
108       400738009       1272469910        2013  tt2294629   
77        417719760       1305772799        2018  tt4881806   
27        700059566       1348258224        2018  tt1825683   
95        504014165       1259199706        2017  tt2771200   
293       264624300       1034727750        2017  tt3469046   
28        608581744       1242520711        2018  tt3606756   
96        335061807       1146894640        2018  tt1477834   
290       368065385        975216835        2013  tt1690953   
32        408992272       1215392272        2013  tt1300854   
383       216303339        894985342        2018  tt1727824   
240       404508916        964496193        2017  tt2283362   
67        426525952       1123061550        2019  tt4154664   
2         459005868       1403013963        2015  tt2395427   
291       368384330        886750534        2016  tt2709768   
16        225764765       1234846267        2017  tt4630562   

                     original_title  runtime_minutes  \
3            Avengers: Infinity War            149.0   
23                   Jurassic World            124.0   
46                    Furious Seven            137.0   
18                     The Avengers            143.0   
301                         Minions             91.0   
108                          Frozen            102.0   
77   Jurassic World: Fallen Kingdom            128.0   
27                    Black Panther            134.0   
95             Beauty and the Beast            129.0   
293                 Despicable Me 3             89.0   
28                    Incredibles 2            118.0   
96                          Aquaman            143.0   
290                 Despicable Me 2             98.0   
32                   Iron Man Three            130.0   
383               Bohemian Rhapsody            134.0   
240  Jumanji: Welcome to the Jungle            119.0   
67                   Captain Marvel            123.0   
2           Avengers: Age of Ultron            141.0   
291         The Secr

We decide to sort the dataset by the profit values, with greatest profit being at the top.

In [113]:
df_budget_merge.sort_values(by = ['profit'], axis = 0, ascending = False, inplace = True)

### Using Genres

Next we check what type of genres are in each row.

In [120]:
df_budget_merge.genres.value_counts()

Adventure,Animation,Comedy    67
Action,Adventure,Sci-Fi       51
Comedy,Drama,Romance          48
Comedy                        48
Drama                         43
                              ..
Biography,Family,Sport         1
Adventure,Comedy,Romance       1
Comedy,Horror,Sci-Fi           1
Comedy,Horror,Mystery          1
Drama,Music,Musical            1
Name: genres, Length: 214, dtype: int64

We decide to use only certain common genres instead of the 214 possible collection of genres.

In [126]:
df_budget_merge['comedy_id'] = [1 if 'Comedy' in x
                                else 0 
                                for x in df_budget_merge['genres']]
df_budget_merge['drama_id'] = [1 if 'Drama' in x
                                else 0 
                                for x in df_budget_merge['genres']]
df_budget_merge['action_id'] = [1 if 'Action' in x
                                else 0 
                                for x in df_budget_merge['genres']]
df_budget_merge.head()

id  release_date                   movie  production_budget  \
3     7  Apr 27, 2018  Avengers: Infinity War          300000000   
23   34  Jun 12, 2015          Jurassic World          215000000   
46   67   Apr 3, 2015               Furious 7          190000000   
18   27   May 4, 2012            The Avengers          225000000   
301  73  Jul 10, 2015                 Minions           74000000   

     domestic_gross  worldwide_gross  start_year     tconst  \
3         678815482       2048134200        2018  tt4154756   
23        652270625       1648854864        2015  tt0369610   
46        353007020       1518722794        2015  tt2820852   
18        623279547       1517935897        2012  tt0848228   
301       336045770       1160336173        2015  tt2293640   

             original_title  runtime_minutes                      genres  \
3    Avengers: Infinity War            149.0     Action,Adventure,Sci-Fi   
23           Jurassic World            124.0     Action,Adventure,Sci-Fi   
46            Furious Seven            137.0       Action,Crime,Thriller   
18             The Avengers            143.0     Action,Adventure,Sci-Fi   
301                 Minions             91.0  Adventure,Animation,Comedy   

     advertisement_budget      profit  comedy_id  drama_id  action_id  
3               300000000  1448134200          0         0          1  
23              215000000  1218854864          0         0          1  
46              190000000  1138722794          0         0          1  
18              225000000  1067935897          0         0          1  
301              74000000  1012336173          1         0          0

We create a column to have a better understanding of the genres in each movie.

In [132]:
df_budget_merge['genre_tuple'] = list(zip(df_budget_merge['comedy_id'], df_budget_merge['drama_id'], df_budget_merge['action_id']))

WE next check to see which movies are missing from our previous genre categories.

In [134]:
df_budget_merge.loc[df_budget_merge['genre_tuple'] == (0, 0, 0)]

id  release_date                                movie  production_budget  \
95   35  Mar 17, 2017                 Beauty and the Beast          160000000   
558  24   Sep 8, 2017                                   It           35000000   
33   51   Mar 5, 2010                  Alice in Wonderland          200000000   
13   19  Dec 14, 2012    The Hobbit: An Unexpected Journey          250000000   
14   21  Dec 13, 2013  The Hobbit: The Desolation of Smaug          250000000   
..   ..           ...                                  ...                ...   
134  17   Jul 1, 2016                              The BFG          140000000   
338  16  Aug 22, 2014         Sin City: A Dame to Kill For           65000000   
142  35   Nov 2, 2018   The Nutcracker and the Four Realms          132900000   
45   66   Mar 1, 2013                Jack the Giant Slayer          195000000   
124  94  Mar 11, 2011                      Mars Needs Moms          150000000   

     domestic_gross  worldwide_gross  start_year     tconst  \
95        504014165       1259199706        2017  tt2771200   
558       327481748        697457969        2017  tt1396484   
33        334191110       1025491110        2010  tt1014759   
13        303003568       1017003568        2012  tt0903624   
14        258366855        960366855        2013  tt1170358   
..              ...              ...         ...        ...   
134        55483770        199646680        2016  tt3691740   
338        13757804         40650842        2014  tt0458481   
142        54858851        170293899        2018  tt5523010   
45         65187603        197687603        2013  tt1351685   
124        21392758         39549758        2011  tt1305591   

                          original_title  runtime_minutes  \
95                  Beauty and the Beast            129.0   
558                                   It            135.0   
33                   Alice in Wonderland            108.0   
13     The Hobbit: An Unexpected Journey            169.0   
14   The Hobbit: The Desolation of Smaug            161.0   
..                                   ...              ...   
134                              The BFG            117.0   
338         Sin City: A Dame to Kill For            102.0   
142   The Nutcracker and the Four Realms             99.0   
45                 Jack the Giant Slayer            114.0   
124                      Mars Needs Moms             88.0   

                         genres  advertisement_budget     profit  comedy_id  \
95       Family,Fantasy,Musical             160000000  939199706          0   
558             Horror,Thriller              35000000  627457969          0   
33     Adventure,Family,Fantasy             200000000  625491110          0   
13     Adventure,Family,Fantasy             250000000  517003568          0   
14            Adventure,Fantasy             250000000  460366855          0   
..                          ...                   ...        ...        ...   
134    Adventure,Family,Fantasy             140000000  -80353320          0   
338              Crime,Thriller              65000000  -89349158          0   
142    Adventure,Family,Fantasy             132900000  -95506101          0   
45            Adventure,Fantasy             195000000 -192312397          0   
124  Adventure,Animation,Family             150000000 -260450242          0   

     drama_id  action_id genre_tuple  
95          0          0   (0, 0, 0)  
558         0          0   (0, 0, 0)  
33          0          0   (0, 0, 0)  
13          0          0   (0, 0, 0)  
14          0          0   (0, 0, 0)  
..        ...        ...         ...  
134         0          0   (0, 0, 0)  
338         0          0   (0, 0, 0)  
142         0          0   (0, 0, 0)  
45          0          0   (0, 0, 0)  
124         0          0   (0, 0, 0)  

[150 rows x 17 columns]

In [135]:
df_budget_merge.describe()

id  production_budget  domestic_gross  worldwide_gross  \
count  1365.000000       1.365000e+03    1.365000e+03     1.365000e+03   
mean     50.795604       4.811613e+07    6.117461e+07     1.533262e+08   
std      28.458309       5.720284e+07    8.684681e+07     2.401767e+08   
min       1.000000       2.500000e+04    0.000000e+00     2.600000e+01   
25%      26.000000       1.000000e+07    6.810754e+06     1.536230e+07   
50%      51.000000       2.600000e+07    3.307827e+07     6.259916e+07   
75%      75.000000       6.000000e+07    7.426203e+07     1.735676e+08   
max     100.000000       4.106000e+08    7.000596e+08     2.048134e+09   

        start_year  runtime_minutes  advertisement_budget        profit  \
count  1365.000000      1359.000000          1.365000e+03  1.365000e+03   
mean   2013.866667       108.706402          4.811613e+07  5.709391e+07   
std       2.607475        17.721580          5.720284e+07  1.671892e+08   
min    2010.000000         5.000000          2.500000e+04 -5.502376e+08   
25%    2012.000000        96.000000          1.000000e+07 -1.405621e+07   
50%    2014.000000       106.000000          2.600000e+07  5.885836e+06   
75%    2016.000000       119.000000          6.000000e+07  6.404180e+07   
max    2019.000000       180.000000          4.106000e+08  1.448134e+09   

         comedy_id     drama_id    action_id  
count  1365.000000  1365.000000  1365.000000  
mean      0.353114     0.486447     0.306227  
std       0.478113     0.499999     0.461094  
min       0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000  
75%       1.000000     1.000000     1.000000  
max       1.000000     1.000000     1.000000

In [18]:
df9= pd.read_csv('./data/zippedData/rt.movie_info.tsv', sep = '\t')
df9.head()

id                                           synopsis rating  \
0   1  This gritty, fast-paced, and innovative police...      R   
1   3  New York City, not-too-distant-future: Eric Pa...      R   
2   5  Illeana Douglas delivers a superb performance ...      R   
3   6  Michael Douglas runs afoul of a treacherous su...      R   
4   7                                                NaN     NR   

                                 genre          director  \
0  Action and Adventure|Classics|Drama  William Friedkin   
1    Drama|Science Fiction and Fantasy  David Cronenberg   
2    Drama|Musical and Performing Arts    Allison Anders   
3           Drama|Mystery and Suspense    Barry Levinson   
4                        Drama|Romance    Rodney Bennett   

                            writer  theater_date      dvd_date currency  \
0                   Ernest Tidyman   Oct 9, 1971  Sep 25, 2001      NaN   
1     David Cronenberg|Don DeLillo  Aug 17, 2012   Jan 1, 2013        $   
2                   Allison Anders  Sep 13, 1996  Apr 18, 2000      NaN   
3  Paul Attanasio|Michael Crichton   Dec 9, 1994  Aug 27, 1997      NaN   
4                     Giles Cooper           NaN           NaN      NaN   

  box_office      runtime             studio  
0        NaN  104 minutes                NaN  
1    600,000  108 minutes  Entertainment One  
2        NaN  116 minutes                NaN  
3        NaN  128 minutes                NaN  
4        NaN  200 minutes                NaN

In [20]:
df10= pd.read_csv('./data/zippedData/rt.reviews.tsv', sep = '\t', encoding = 'latin-1')
df10.head()

id                                             review rating   fresh  \
0   3  A distinctly gallows take on contemporary fina...    3/5   fresh   
1   3  It's an allegory in search of a meaning that n...    NaN  rotten   
2   3  ... life lived in a bubble in financial dealin...    NaN   fresh   
3   3  Continuing along a line introduced in last yea...    NaN   fresh   
4   3             ... a perverse twist on neorealism...     NaN   fresh   

           critic  top_critic         publisher               date  
0      PJ Nabarro           0   Patrick Nabarro  November 10, 2018  
1  Annalee Newitz           0           io9.com       May 23, 2018  
2    Sean Axmaker           0  Stream on Demand    January 4, 2018  
3   Daniel Kasman           0              MUBI  November 16, 2017  
4             NaN           0      Cinema Scope   October 12, 2017

In [ ]:
df11= pd.read_csv('./data/zippedData/bom.movie_gross.csv')
df11.head()